services/file-retriever/src/services/dsnFetcher.ts

```diff
@@ -141,6 +146,123 @@ const fetchObjects = async (objects: ObjectMapping[]) => {
   )
 }
 
+const getNodesForPartialRetrieval = async (
+  chunks: ExtendedIPLDMetadata[],
+  byteRange: ByteRange,
+): Promise<{
+  nodes: string[]
+  firstNodeFileOffset: number
+}> => {
+  let accumulatedLength = 0
+  const nodeRange: [number | null, number | null] = [null, null]
+  let firstNodeFileOffset: number | undefined
+  let i = 0
+
+  logger.debug(
+    `getNodesForPartialRetrieval called (byteRange=[${byteRange[0]}, ${byteRange[1] ?? 'EOF'}])`,
+  )
+
+  // Searchs for the first node that contains the byte range
+  while (nodeRange[0] === null && i < chunks.length) {
+    const chunk = chunks[i]
+    const chunkSize = Number((chunk.size ?? 0).valueOf())
+    // [accumulatedLength, accumulatedLength + chunkSize) // is the range of the chunk
+    if (
+      byteRange[0] >= accumulatedLength &&
+      byteRange[0] < accumulatedLength + chunkSize
+    ) {
+      nodeRange[0] = i
+      firstNodeFileOffset = accumulatedLength
+    } else {
+      accumulatedLength += chunkSize
+      i++
+    }
+  }
+
+  // Searchs for the last node that contains the byte range
+  // unless the byte range is the last byte of the file
+  if (byteRange[1]) {
+    while (nodeRange[1] === null && i < chunks.length) {
+      const chunk = chunks[i]
+      const chunkSize = Number((chunk.size ?? 0).valueOf())
+      if (
+        byteRange[1] >= accumulatedLength &&
+        byteRange[1] < accumulatedLength + chunkSize
+      ) {
+        nodeRange[1] = i
+      }
+      accumulatedLength += chunkSize
+      i++
+    }
+  }
+
+  if (nodeRange[0] == null) {
+    throw new Error('Byte range not found')
+  }
+
+  const nodes = chunks
+    .slice(nodeRange[0], nodeRange[1] === null ? undefined : nodeRange[1] + 1)
+    .map((e) => e.cid)
+
+  return {
+    nodes,
+    firstNodeFileOffset: firstNodeFileOffset ?? 0,
+  }
+}
+
+const fetchFileAsStreamWithByteRange = async (
+  cid: string,
+  byteRange: ByteRange,
+): Promise<Readable> => {
+  const chunks = await dsnFetcher.getFileChunks(cid)
+  const { nodes, firstNodeFileOffset } = await getNodesForPartialRetrieval(
+    chunks,
+    byteRange,
+  )
+
+  logger.debug(
+    `getNodesForPartialRetrieval called (byteRange=[${byteRange[0]}, ${byteRange[1] ?? 'EOF'}]) nodes=${JSON.stringify(nodes)} firstNodeFileOffset=${firstNodeFileOffset}`,
+  )
+
+  // We pass all the chunks to the fetchNode function
+  // So that we can fetch all the nodes within the same piece
+  // in one go
```

Comment:
Some chunks can cross piece boundaries, so it's even better than that. It fetches all the nodes within a series of pieces, by only downloading and validating each piece once.

(This doesn't need any changes, just double-checking everyone's understanding here.)